pkg/config/sync.go

```diff
@@ -0,0 +1,207 @@
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	crclient "sigs.k8s.io/controller-runtime/pkg/client"
+
+	controller "github.com/devfile/devworkspace-operator/apis/controller/v1alpha1"
+	"github.com/devfile/devworkspace-operator/pkg/infrastructure"
+)
+
+const (
+	OperatorConfigName     = "devworkspace-operator-config"
+	openShiftTestRouteName = "devworkspace-controller-test-route"
+)
+
+var (
+	Routing         *controller.RoutingConfig
+	Workspace       *controller.WorkspaceConfig
+	internalConfig  *controller.OperatorConfiguration
+	configMutex     sync.Mutex
+	configNamespace string
+)
+
+func SetConfigForTesting(config *controller.OperatorConfiguration) {
+	configMutex.Lock()
+	defer configMutex.Unlock()
+	internalConfig = DefaultConfig.DeepCopy()
+	mergeConfig(config, internalConfig)
+	updatePublicConfig()
+}
+
+func SetupControllerConfig(client crclient.Client) error {
+	if internalConfig != nil {
+		return fmt.Errorf("internal controller configuration is already set up")
+	}
+	internalConfig = &controller.OperatorConfiguration{}
+	namespace, err := infrastructure.GetNamespace()
+	if err != nil {
+		return err
+	}
+	configNamespace = namespace
+	config, err := getClusterConfig(configNamespace, client)
+	if err != nil {
+		return err
+	}
+	if config == nil {
+		internalConfig = DefaultConfig.DeepCopy()
+		updatePublicConfig()
+	} else {
+		syncConfigFrom(config)
+	}
+	if internalConfig.Routing.ClusterHostSuffix == "" && infrastructure.IsOpenShift() {
+		if err := fillOpenShiftRouteSuffix(client); err != nil {
+			return err
+		}
+		updatePublicConfig()
+	}
+	return nil
+}
+
+func ExperimentalFeaturesEnabled() bool {
+	if internalConfig.EnableExperimentalFeatures == nil {
+		return false
+	}
+	return *internalConfig.EnableExperimentalFeatures
+}
+
+func getClusterConfig(namespace string, client crclient.Client) (*controller.DevWorkspaceOperatorConfig, error) {
+	clusterConfig := &controller.DevWorkspaceOperatorConfig{}
+	if err := client.Get(context.Background(), types.NamespacedName{Name: OperatorConfigName, Namespace: namespace}, clusterConfig); err != nil {
+		if k8sErrors.IsNotFound(err) {
+			return nil, nil
+		}
+		return nil, err
+	}
+	return clusterConfig, nil
+}
+
+func syncConfigFrom(newConfig *controller.DevWorkspaceOperatorConfig) {
+	if newConfig == nil || newConfig.Name != OperatorConfigName || newConfig.Namespace != configNamespace {
+		return
+	}
+	configMutex.Lock()
+	defer configMutex.Unlock()
+	internalConfig = DefaultConfig.DeepCopy()
+	mergeConfig(newConfig.Config, internalConfig)
+	updatePublicConfig()
+}
+
+func restoreDefaultConfig() {
+	configMutex.Lock()
+	defer configMutex.Unlock()
+	internalConfig = DefaultConfig.DeepCopy()
+	updatePublicConfig()
+}
+
+func updatePublicConfig() {
+	Routing = internalConfig.Routing.DeepCopy()
+	Workspace = internalConfig.Workspace.DeepCopy()
+}
+
+func fillOpenShiftRouteSuffix(client crclient.Client) error {
+	if !infrastructure.IsOpenShift() {
```

Comment:
This check is done twice:
```go
...
   if internalConfig.Routing.ClusterHostSuffix == "" && infrastructure.IsOpenShift() {
		if err := fillOpenShiftRouteSuffix(client); err != nil {
			return err
		}
		updatePublicConfig()
	}
...

func fillOpenShiftRouteSuffix(client crclient.Client) error {
	if !infrastructure.IsOpenShift() {	
```
This suggests something is wrong.

Maybe it makes sense to refactor it in the following way:
```go
    if internalConfig.Routing.ClusterHostSuffix == "" && infrastructure.IsOpenShift() {
        routeSuffix, err := discoverRouteSuffix(client);
		err != nil {
			return err
		}
		internalConfig.Routing.ClusterHostSuffix = clusterHostSuffix
		// Set routing suffix in default config as well to ensure value is persisted across config changes
		DefaultConfig.Routing.ClusterHostSuffix = clusterHostSuffix
		updatePublicConfig()
	}
```

Reply:
Reworked as suggested. I'm only checking `IsOpenShift()` once now, and `discoverRouteSuffix` returns `"", nil` when on Kubernetes.